### DISCUSSION TITLE
Filter MySQL Table by Current User Email

### DISCUSSION THREAD

**Discussion by mmalessa80** [Budibase/budibase#6198] [Help]
I am attempting to filter a MySQL table by the current users email against a column that is an email field. When saving the filter, the table reloads but without any records even though there is an entry for the current user's email. When test running the application, the filter is not applied as it returns all records on the client and when returning to editor mode the filter no longer is there, as in it was automatically erased. Is it not possible to link a table filter to a binding? What would be a possible work around?

![image](https://user-images.githubusercontent.com/106758851/172206502-f4f20781-7b54-499e-936b-741e714501a5.png)

**Comment by melohagan** [comment]
Hi @mmalessa80 

I think you just need to click on the `Value` dropdown and select `Binding`

**Reply by mmalessa80** [reply]
It appears that I do not have that option. This is in the actual Table filter constraint rather than a query. Not sure if that makes a difference.

![image](https://user-images.githubusercontent.com/106758851/172211804-d95eb0c0-5b43-44a7-aa69-a6cedf03265f.png)

**Reply by melohagan** [reply]
Oh - that's odd, it should be there. Looks like a bug to me. What data type is **Reporter_Email** in your MySQL table?

**Reply by mmalessa80** [reply]
Should be just text. The column is defined as VARCHAR(255).

**Reply by melohagan** [reply]
Hmm, I can't recreate this. What version of budibase is your app using? 
It may be helpful to see your app export if you want to export it to support@budibase.com

**Reply by mmalessa80** [reply]
It is version 1.0.195. Which supposedly is the latest version.

**Reply by melohagan** [reply]
Is there an option to rollback the version to the previous, just in-case there's been a recent change?

**Reply by mmalessa80** [reply]
I was able to revert back to version 1.0.194, however I still did not have an option to choose a binding instead of a value.

**Reply by mmalessa80** [reply]
I came across this topic https://github.com/Budibase/budibase/discussions/3684 and see what you are referring to as having the option to change to a binding. However I just don't have that available. Would there be a reason why that is?

**Reply by melohagan** [reply]
I can't think of any reason tbh. It may help to see the app export.

**Reply by mmalessa80** [reply]
Is the export just the text export and is it ok to just attach it to a reply or copy the text? Just to test, I started a new application from scratch rather than a template, added MySQL connection and the tables in my database, then went to the same table and tried doing a filter and still only have 'Value' as an option.

What I'm really trying to accomplish is prefilter the data returned from my database, like a security filter. I have a few tables that should be filtered by the current user and then I wanted to filter other tables based on the results of these tables. It appears that MySQL tables cannot be filtered simply by a select statement, so is my only option to use Queries and only return the query results as a data provider within the client? Does that prevent a user from seeing the whole dataset though?

**Reply by melohagan** [reply]
In your app list, click on the ellipsis and select export app. You can then attach that and send it to support@budibase.com

<img width="1195" alt="Screenshot 2022-06-07 at 21 29 45" src="https://user-images.githubusercontent.com/101575380/172476553-d9f7ce80-0333-4657-b3c2-f71a72b4fab3.png">

Also, if you send the CREATE script for the table you're trying to filter, I can use my own datasource to test with.

**Reply by mmalessa80** [reply]
I sent the email. Thanks for having a closer look.

**Reply by melohagan** [reply]
So I did try out the filter on the latest versions 1.0.187 and 1.0.196 and it worked for me in a Data Provider and Table Block for the *Reporter_Email* column.

There was no existing table in your app export that I could find with the filter already set for that column. Had you saved the value only filter?

**Reply by mmalessa80** [reply]
Just wanted to check, but you are trying this in the 'Design', i.e. client side design? Maybe that is where the confusion lies, since I am attempting to get this accomplished in the 'Data' block. But it appears when setting a filter there and saving it, only filters the data while on that screen. Once I select a different table and then return to the ReporterInfo table the filter no longer exists. Therefore, when I do an export that filter would probably not exists any longer in the export.

![image](https://user-images.githubusercontent.com/106758851/172624792-f8337593-0777-45e0-a3a1-b8c6515cd2ea.png)

**Comment by melohagan** [comment]
Hey @mmalessa80 

Thank you for clarifying :)

I actually don't think I've used that filter before! The reason it doesn't allow bindings is that the filter applied at the data level does not impact the design or the app in anyway.
In fact whenever you apply a filter on the data screen and click onto the design tab, the filter is cleared. It's simply a tool for finding data as a developer - it has nothing to do with the app.

If you want to filter on the logged in users email, the use a <a href="https://docs.budibase.com/docs/searchfilter-data#static-filtering">static filter</a> with the `{{ Current User.email }}` binding.

Without trying to confuse matters, I have raised an issue to allow bindings through to internal table views filters: https://github.com/Budibase/budibase/issues/6108
<a href="https://docs.budibase.com/docs/views">Views</a> are a special case however, and are more of a counterpart to external table <a href="https://docs.budibase.com/docs/custom-queries">custom queries</a> .

**Reply by mmalessa80** [reply]
Thank you for the clarification. I have actually set up a query like this

![image](https://user-images.githubusercontent.com/106758851/172632367-06afa909-0093-4be3-8cbd-ffdbc051bdf1.png)

Although when I link a data provider on the client to this query, I do get filtered results, however the filtered results do not reflect the correct result. It actually returns 5 rows with the Reporter_Email field actually being empty.
Furthermore, using query results as data providers I am loosing the ability to access relationships that I set up under the MySQL tables unless I incorporate JOINS in my query or run subqueries based on the results of the main query.

It is possible that the current budibase configuration simply does not align with my current use case, which would implement a filter or SELECT statement at the data table level (server side). My use case would implement returning a single row in the ReporterInfo table, then returning rows in the PermitList table based on a match in my bridge table called ReporterInfo_Has_PermitList. Then I have approximately 8 tables that would get filtered based on the results from the PermitList table also through their respective bridge tables.

**Reply by melohagan** [reply]
You're not quite using the binding correctly. Have a look at this page here: https://docs.budibase.com/docs/custom-queries

You need to click the 'Add Binding' button. Call it say 'UserEmail'.
Then your query would be `... WHERE Reporter_Email = {{ UserEmail }}`

Then in the **design** tab, where you are using the query as a datasource, you would click the cog, and set *UserEmail* to `{{ Current User.email }}`
More here: https://docs.budibase.com/docs/custom-queries#using-query-bindings

Hopefully that makes sense!

**Reply by mmalessa80** [reply]
Ok, I think I got that part working, albeit as I mentioned above using this route I would either need to incorporate joins to get my many-to-many relationships to show up in the query results or add additional (sub)queries based on the results from another. Would this affect form handling as it pertains to relationship pickers also?

I am also wondering if the client side binding can be intercepted this way and altered to give someone access to other data?

Thanks again for your help in clarifying!
